internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala

```diff
@@ -90,33 +90,29 @@ object ClassToAPI {
    *
    * In order to fully fix this issue, the Scala compiler should do the desired behaviour
    * in 2.13.x. 2.12.x is already lost since this fix is binary incompatible. The appropriate
-   * JIRA ticket is here: https://github.com/scala/bug/issues/2034.
+   * issue is here: https://github.com/scala/bug/issues/2034.
    *
-   * @return The canonical name if not null, the full name otherwise.
+   * @return The canonical name if not null, the blank string otherwise.
    */
-  def handleMalformedNameOf(c: Class[_], isRecursive: Boolean = false): String = {
+  def classCanonicalName(c: Class[_]): String = {
     if (c == null) "" // Return nothing if it hits the top-level class
     else {
       val className = c.getName
-      // Adds a `dollar` if it's a recursive call, else nothing
-      val atEnd: String = if (isRecursive) "$" else ""
       try {
         val canonicalName = c.getCanonicalName
         if (canonicalName == null) className
-        else canonicalName + atEnd
+        else canonicalName
       } catch {
         case malformedError: java.lang.InternalError
             if malformedError.getMessage.contains("Malformed class name") =>
           val enclosingClass = c.getEnclosingClass
           val enclosingName = enclosingClass.getName
           val restOfName = c.getName.stripPrefix(enclosingName)
-          handleMalformedNameOf(enclosingClass, true) + restOfName + atEnd
+          classCanonicalName(enclosingClass).stripSuffix("$") + "." + restOfName
```

Comment:
> Zinc assumes that both class and companion class map to the same name. 

I presume you mean for the enclosing classes, given it's not true for top level class pairs (e.g O and C), or leaf class pairs (e.g OO and OC).

> This change, however, creates a clash because two different class files (one with a valid name and another one with an invalid name) will map to the same name. Those don't have any relationship between them, as in the case of class and companion class.

The relationship between OOO and OCO is a class and companion class relationship at the middle tier.

(btw I've said "relationship" enough time that at this point I feel like I'm commenting about a TV soap opera.. "Jorge, why can't OCO love OOO, just like he loves CCO?")